### PR TITLE
Add a grafana dashboard

### PR DESCRIPTION
**Pull request by rbarry82**
And fix the prometheus tests. This also moves it to Juju topology

### CODE REVIEW

**Review by dstathis — APPROVED** [review]

File: src/charm.py [hunk 1]
```diff
@@ -103,9 +106,31 @@ def __init__(self, *args):
...
+        self.dashboard_consumer = GrafanaDashboardConsumer(
+            charm=self, name="grafana-dashboard", consumes={"Grafana": ">=2.0.0"}
+        )
+        self.framework.observe(
+            self.dashboard_consumer.on.dashboard_status_changed,
+            self._on_dashboard_status_changed,
+        )
+
+        self._dashboard_valid = False
+        self._defer_dashboard_send = False
```
Comment (sed-i) [inline src/charm.py]: at first glance it looks like these should be _stored..?

File: src/charm.py [hunk 2]
```diff
@@ -127,9 +152,55 @@ def on_database_joined(self, event):
         self.provider.update_port("database", self.model.config["port"])
         self.provider.update_address("database", self._bind_address())
 
+    def _on_dashboard_joined(self, event):
+        if not self.unit.is_leader():
+            return
```
Comment (sed-i) [inline src/charm.py:157]: Is the leader guard needed? If yes. then is it needed so far upstream?

File: src/charm.py (hunk 1 above)
Comment (rbarry82) [reply src/charm.py]: They don’t need to survive a charm failure/restart, so no (it’s actually somewhat better for the logic to reset), but please hold off on a review.

I’m refactoring this sort of logic back into the library to make it reusable.

File: src/charm.py [hunk 3]
```diff
@@ -127,9 +152,55 @@ def on_database_joined(self, event):
...
+        # when monitoring is joined will cause the check for the number
+        # of Prometheus units in the dashboard consumer library to raise
+        # an IndexError, since it hasn't finished yet. Delay...
+        if self._defer_dashboard_send:
+            self._defer_dashboard_send = False
+            event.defer()
+            return
+
+        dashboard_tmpl = open(
+            os.path.join(sys.path[0], "dashboard.json.tmpl"), "r"
```
Comment (sed-i) [inline src/charm.py:166]: I always [use inspect](https://stackoverflow.com/a/6209894) instead of `sys.path[0]`. Haven't checked recent advances.

Comment (dstathis) [reply src/charm.py:166]: Perhaps we should be using resources for the dashboard?

**Comment by dstathis** [comment]
By the way, I'm working on a similar PR. I'll wait until this is in then I can merge it in to my fork.

File: src/charm.py (hunk 3 above)
Comment (rbarry82) [reply src/charm.py:166]: We know for sure that the charm runs from the same path, there is no chance of a reexec somewhere else, and the dashboard will be colocated. Using `inspect` is really overkill.

I'm not against making this a resource, except that it makes the process of publishing it that much more complicated for what is, really, a simple text file.

File: src/charm.py [hunk 4]
```diff
@@ -127,9 +152,41 @@ def on_database_joined(self, event):
...
+            self._dashboard_valid = True
+            self.unit.status = ActiveStatus()
+        elif event.error_message:
+            self._dashboard_valid = False
+            self.unit.status = BlockedStatus(event.error_message)
+
     @status_catcher
-    def on_monitoring_joined(self, _):
-        self._setup_monitoring()
+    def on_monitoring_joined(self, event):
```
Comment (sed-i) [inline src/charm.py:188]: Type annotation for event would be helpful. Even if it is EventBase.

File: src/charm.py [hunk 5]
```diff
@@ -156,24 +212,24 @@ def _setup_monitoring(self):
...
-                    container.push(ENV_PATH, "\n".join(cassandra_env))
-                    restart(container)
-                    break
+        if len(self.model.get_relation("monitoring").units) == 0:
+            try:
+                container = self.unit.get_container("cassandra")
+                cassandra_env = container.pull(ENV_PATH).readlines()
+                for line in cassandra_env:
+                    if "jmx_prometheus_javaagent" in line:
+                        cassandra_env.remove(line)
```
Comment (sed-i) [inline src/charm.py:225]: [answers on SO](https://stackoverflow.com/questions/1207406/how-to-remove-items-from-a-list-while-iterating#1207461) seem to suggest that it is not valid to update a list like that while iterating.
Indeed, you can try:
```
>>> l = [1,2,3,4]
>>> for item in l:
...     l.remove(item)
>>> print(l)
>>> l
[2, 4]
```
iterator invalidation?

Comment (sed-i) [reply src/charm.py:225]: in the context of this code it doesn't matter because of the `break` though

File: src/charm.py (hunk 3 above)
Comment (sed-i) [reply src/charm.py:166]: a 7251-line long simple text file :)

**Review by sed-i — APPROVED** [review]

**Review by Abuelodelanada — APPROVED** [review]
LGTM

**Pull request merged**